README.md
# FAQ 
**Fulgence** is a precision, highly customizable, text-based, REST API friendly railroad simulator made of three core components: 
- mp.py, the train engine simulator itself. Sims are complex beasts, you should check out our *tutorials below*
- room.py, the control room (showing enroute trains progress). **A demo of the control room is [available here](http://fulgence.lovethosetrains.com/controlRoom.html)** 
- tr.py, the traffic manager (the brain, providing path information to the simulator). Under development.

Here is a sample of what you can get in the control room:

![alt text](https://github.com/freevariable/Fulgence/blob/master/capture.png "Paris metro demo")

## Main features
Fulgence will let you perform various tasks:
- [x] define realistic (eg: Paris Metro Line 1) or fancy routes (eg: The Polar Comet) with grades and curves
- [x] define realistic or fancy schedules on these routes
- [x] define your own rolling stock (EMUs and steam engines)
- [x] save sim, restore and resume at a later time
- [x] get live simulation JSON data through a REST API (see below supported verbs)
- [x] run schedules in real-time or accelerated time
- [ ] scale train services according to passengers peaks (TRAFFIC MANAGER feature)
- [x] place aspect signals for train separation and route branches
- [ ] place aspect signals for trains overtaking and single track operations (TRAFFIC MANAGER feature)
- [x] set headway timers for trains separation
- [ ] place electronic signals for trains separation and route branches 
- [x] calculate live power, water and coal consumption
- [x] calculate wear and tear
- [ ] identify servicing thresholds for brakes, vapor engine and moving parts. Implement breakdowns!
- [x] oversee all trains progress in the control room of your route
- [ ] update train schedules in realtime depending on time of the day/congestion (TRAFFIC MANAGER feature)

### What works as of today (build ALPHA TWO)?
Since the simulator is in ALPHA, only a subset of features are currently usable:
- you may run the sim on the *ParisLine1* and *LondonCentral* routes, with any schedule (the predefined ones or your own custom schedules)
- you may run the control room for the *ParisLine1* and *LondonCentral* routes
- you may run vanilla or custom steam engines in the *PolarComet* route until it runs out of resources (coal or water)
- you may run vanilla or custom steam engines in *TheCorkScrew* route, upward or downward segments.
- four API verbs are currently supported: v1/list/schedules, v1/describe/schedule/*scheduleName*, v1/save/*saveName*/*time* and v1/describe/status
- you may save a running sim, stop it and resume later one
- you may NOT run the traffic manager as of now (this is slated for ALPHA THREE)

The steam engine is giving promising results!! Here are the characteristic curves of a 147tons Atlantic (including tender) with two cylinders (not compounded) and a 250t payload:

![alt text](https://github.com/freevariable/Fulgence/blob/master/250t.png "Atlantic")

![alt text](https://github.com/freevariable/Fulgence/blob/master/250t.a.png "Atlantic")

## Notes for route designers
### Layout
- routes will support both on-track signals and electronic signals (only on-track for now)
- Routes are located in (Fulgence folder)/(routeName)/ 
- The global parameters such as gauge, electrification, imperial or metric units... are located in (routeName)/routeConfig.txt. Only units must be set for now, the other parameters are optional
- Routes are made of segments located in (routeName)/segments/(segmentName)/. For example: ParisLine1/segments/WestboundMain/ for the default route
- A segment is a set of contiguous blocks where each block is under control of a signal
- Blocks are not explicitely delimited or managed in Fulgence. In fact, signals define blocks.
- Blocks are unidirectional, except the ones protected by a **reversing signal** 
- Consequently, segments are unidirectional. It means that single track routes are not supported yet.
- In each segment folder, you need to describe the location of signals (SIGs), stations (STAs), speed limits (TIVs), radius of curvature (CRVs) and grades (GRDs). 
- All route and schedule data are kept in simple text files. Lines beginning with a hash are ignored, as one would expect.
- Schedules are kept in the (routeName)/schedules directory. You may name them whatever.txt The default shedule is default.txt
- Services are kept in (routeName)/services.txt This file is **optional**. Use it only if you have branches on your line to provide pathfinding information.
- Markers are not supported yet. In the pipe: tunnels (provide weather protection), country boundaries (with imperial/metric unit changes), points of interest (landmarks), areas of interest, platform names, substations (for line sectioning).
- Weather is not supported yet
- Steam engines fully replenish water and coal at all stations on their way

### Signal types
The following on-track signals are implemented:
- [x]  *Type 1*: this is the usual 3-aspect signal. The possible states are: VL (green, all clear), A (yellow, prepare to stop at next signal) and C (red, impassable stop).
- [x] *Type 2*: this is a buffer signal that allows reversing to a segment which is different from the origin segment. As far as the origin segment is concerned, its only possible state is C (red, impassable) and it is preceded by a type 3 signal. As for the first signal in the reversed direction, it is controlled by a type 5 signal.
- [x] *Type 3*: this is a 2-aspect signal (A and C). It  must always **precede** a type 2, as its state depends on the switch position in the upcoming junction.
- [x] *Type 4D*: this is a diverging junction signal, used to stitch segments together. Both legs of the junction see trains **moving in the same direction**. (As opposed to the second type of diverging junction described below). Type 4D must **always be preceded** by a type 1.
- [x] *Type 4C*: this is a converging junction signal, used to stitch segments together. Both legs of the junction see trains **moving in the same direction**. It must **always be preceded** by a type 6 in both its legs.
- [x] *Type 5*: this is a diverging junction signal for trains coming from a reversing block. It must always **be preceded** by a type 2 signal. One leg of the junction is for trains **coming from** the forward direction, the other leg is for trains **going to** the reverse direction.
- [x] *Type 6*: this is a 3-aspect signal **always preceding** a type 4C signal. So there are two such type 6 signals for any given type 4C: one in each leg. Both type 6 sort of compete to control the switch position in the 4C block. 

Junction signals can manage only two legs, no less, no more. One leg is the main segment (left or right), the other one to the diverging/converging segment (left or right).Junction signals must have a unique name in both segments so that the engine may perform the segments stitching properly.

### Signals placement per segment type
- *Main branches* must start by a type 2 (succeeded by a type 5) and terminate by a type 2 (preceded by a type 3) signal. See for example: LondonCentral/segments/Epping/SIGs.txt
- *Sidings must* start with a type 4D signal and terminate with a type 4C signal (preceded by a type 6 signal)
- *Secondary branches* diverging from a main branch must start with a 4C signal and terminate with a type 2 signal (preceded by a type 3). See for exemple: LondonCentral/segments/Hainault/SIGs.txt
- *Exit to a garage* must start with a 4C signal and terminate with a type 1 signal. See for exemple: ParisLine1/segments/FontenayExit/SIGs.txt
- *Secondary branches* converging towards a main branch must start with a type 2 signal (succeeded by a type 5) and terminate with a type 4C signal (preceded by a type 6). See for exemple: LondonCentral/segments/WestActon/SIGs.txt
- *Entry from a garage* must start with a type 2 signal (succeeded by a type 5) and terminate by a type 4C (preceded by a type 6 signal)

## Installation 
sudo apt-get update

sudo apt-get install -y redis-server python curl python-redis python-flask

Then: clone Fulgence from GitHub and... voila!

## Tutorials
### Tutorial 1: run the small schedule of the LondonCentral route for 1 hour 
- mp.py --route=LondonCentral --schedule=tr.txt --duration=3600

Since mp.py is a daemon, from that point nothing will seem to happen but the sim will actually have started. To know what is going on, you have two options: either run the control room, or use the API. This is the purpose of the next tutorial.

### Tutorial 2: see real time trains progress
In this tutorial, we suppose that you have a web server up and running, with the document root located in /var/www/html

To use the control room, we must *restart the sim in realtime mode*: 
mp.py --realtime --route=LondonCentral --schedule=tr.txt --duration=3600

While the sim is running in background, we then run room.py at regular intervals, say every minute: 
controlRoom/room.py --route=LondonCentral --schedule=tr.txt --segments=Epping,WestRuislip > /var/www/html/controlRoom.html

After every room.py execution, we need to copy the generated files controlRoom/Epping.html and controlRoom/WestRuislip.html to the web server, in /var/www/html

We also need to copy all the svg files located in controlRoom/html/ to the web server:
cd controlRoom/html && cp *.svg /var/www/html/

Point your browser to http://yourwebserver/controlRoom.html and see the trains progress and the stations on the various segments of the line.

You may also want to probe the sim with an API call:
- curl http://127.0.0.1:4999/v1/describe/status
Returns usefull data about the running sim: its ID, the current time, the route name, etc

- curl http://127.0.0.1:4999/v1/list/schedules
This will bring up a JSON list of all trains currently running on the line.

## Simulator (mp.py)
### Save and restore sim
By default, the sim is saved to a file in the *saves* subdirectory every 15 minutes
You may want to schedule a specific sim save under name "myName" at (simulation) time t=55 seconds by issuing:
- curl http://127.0.0.1:4999/v1/save/myName/55

myName will be saved in the *saves* subdirectory.
To restore the sim:
- kill the running mp.py
- restart mp.py with the resume=simName option: ./mp.py --resume=myName 

You may start the sim in accelerated time and resume in realtime or the opposite: all combinations are possible. :)

### Simulator API
- get a status of a runnin sim: curl http://127.0.0.1:4999/v1/describe/status
- use curl to dump the list of currently active schedules on the route: curl http://127.0.0.1:4999/v1/list/schedules
- based on the list of schedules, get live data on a specific schedule: curl http://127.0.0.1:4999/v1/describe/schedule/scheduleName
- save the sim in file *mySaveName.pickle* after 3 minutes (180 seconds) by calling curl http://127.0.0.1:4999/v1/save/mySaveName/180

Exemple on *ParisLine1*, using the default schedule.txt :
curl http://127.0.0.1:4999/v1/describe/schedule/E500

## Control room (controlRoom/room.py)
The control room is displayed as an HTML dashboard by calling *tools/room.py* **after** or **while** you run mp.py in realtime. It will not work otherwise, because it polls redis for live information and redis will be empty.

*room.py* also generates one HTML file per segment you wish to monitor. It is called *segmentName.html*

For stations to appear on the dashboard, they must be succeded by a signal named (station trigram)+(whatever) in (routeName)/segments/(routeSegment)/SIGs.txt

### Control room options
- You must provide the route name using *--route*
- You must provide a segments list using *--segments*  Segment names are separated by a comma
- You must provide a schedule name using *--schedule*

Here are two  examples:

*controlRoom/room.py --route=ParisLine1 --schedule=large.txt --segments=WestboundMain,EastboundMain > dashboard.html* 

*controlRoom/room.py --route=LondonCentral --schedule=default.txt --segments=WestRuislip,Epping,Hainault,Wanstead,Acton,EalingBroadway > dashboard.html* 

The first example will produce three HTML files: dashboard.html, WestboundMain.html and EastboundMain.html

In the second example, seven HTML files will be produced.

## Traffic manager (tr.py)
To be completed. The traffic manager is not avalaible for now.


ReleaseNotes.md
## What's new in build 149
- Finished raw implementation of the downScale feature of the traffic manager (in prep for ALPHA 3)
- Modified ParisLine1 (added switches and signalling to Maillot Workshop) 

## What's new in build 148
- Started implementation of the downScale feature of the traffic manager (in prep for ALPHA 3)

## What's new in build 147
- More traffic manager updates (in prep for ALPHA 3)

## What's new in build 146
- Traffic manager update (in prep for ALPHA 3)

## What's new in build 145
- Improved sim load and resume feature 

## What's new in build 144
- Implemented sim load and resume feature

## What's new in build 143
- Optimized wear and tear tick calculations
- Implemented sim save feature (sim load not implemented yet)
- Two new API verbs: v1/describe/status and /v1/save/saveName/saveTime
- Work on traffic manager

## What's new in build 142
- Implement steam engine coal & water consumption cut off in slopes
- Started implementation of wear and tear

## What's new in build 141 
- Traffic manager improvement and interfacing to the sim via the REST API

## What's new in build 140 
- Continued implementation of the simulator REST API (see README) in preparation for ALPHA 3: traffic manager  
- Fixed bugs in multithreading. Should be fine, now

## What's new in build 139
- Skeleton implementation of the traffic manager, located in directory trafficManager/ 

## What's new in build 138
- Started implementation of the simulator REST API (see README) in preparation for ALPHA 3: traffic manager

## What's new in build 137
- Bug fixing in the simulator, in preparation for ALPHA 3: traffic manager
- Created trafficManager directory

## What's new in build 136
- Bug fixing in steam engine calculations, TheCorkScrew route and the control room

## What's new in build 135
- Buffed AulderKraak Consolidation
- Completed TheCorkSscrew upward and downward segments

## What's new in build 134
- Ended implementation of tracks radius of curvature
- Bug fixing
- Modified ParisLine1 (added switches and signalling to Fontenay Workshop)

## What's new in build 132
- Steam engines copied to locoLib/
- New steam engine: AulderKraak Consolidation
- New (undocumented) route: TheCorkScrew

## What's new in build 131
- Started implementation of tracks radius of curvature

## What's new in build 130
- Added the last missing parameters to the steam engines rollingStock.txt file
- New steam engine: Decapod Est 150-001 (temporarily located in PolarComet/rollingStock.txt)

## What's new in build 129 (ALPHA TWO)
- LondonCentral now stable under medium load (schedules/medium.txt)!
- Added several parameters to the steam engines rollingStock.txt file
- Extended the PolarComet northbound segment to Newhaven

## What's new in build 125
- Another row of bug fixing in preparation for release ALPHA TWO
- Added aspect signals to LondonCentral
- Added speed signals to LondonCentral

## What's new in build 124
- Lots of bug fixing in preparation for release ALPHA TWO

## What's new in build 123
- Bugs fixing
- Imperial units support in Control Room

## What's new in build 122
- Imperial units are now supported for distances and speeds. Units are set in (routeDirectory)/routeConfig.txt
- routeConfig.txt set in PolarComet,ParisLine1, TestTrack and LondonCentral
- finished the LondonCentral route minimum viable signalling to connect the 6 segments.

## What's new in build 121
- Trains are removed from the services list when they reach a non-type 2 signal
- Non-type 2 signals with no successor signals are set to red
- Simulation ends gracefully when services list becomes empty

## What's new in build 120
- Steam engines now refill coal and water at stations
- The average waiting time at stations (in seconds) is now set in rollingStock.txt under the *waitTime* attribute
- Drivers reaction time after sig clearance has been slightly randomized
- Placed LondonCentral stations in STAs.Txt in the six segment folders


tools/profileGen.py
#!/usr/bin/python

INITHEIGHT=30.0    # in m
FLATENESS=0.64      #chance of flat terrain
LENGTH=159.4   #in km
MAXGRADE=0.002  #in radian (or meter per meter)
MINSECTION=130.0   #in m
MAXSECTION=MINSECTION*8.3    #in m

import random
import sys

x=0.0
z=30.93
curOrderAvg=0.70
secondOrder={}
secondOrder[0]={'x':0.0,'avg':0.7,'flat':0.64,'max':0.003}
secondOrder[1]={'x':430000.0,'avg':0.03,'flat':0.36,'max':0.006}
secondOrder[2]={'x':490000.0,'avg':0.50,'flat':0.64,'max':0.004}
secondOrder[3]={'x':9999990000.0,'avg':0.50,'flat':0.64,'max':0.004}
oldGradient=0.0

def section(initTrend):
  global x
  global z
  global oldGradient
# step 1: determine trend
  trendRand=random.uniform(0.0,1.0)
  slopeRand=random.uniform(0.0,1.0)
  if (initTrend=='0'):
    trend='='
  elif (initTrend=='='):
    if trendRand<=curOrderFlat:
      trend='='
    else:
      if slopeRand<curOrderAvg:
        trend='-'
      else:
        trend='+' 
  elif (initTrend=='-'):
    if trendRand<=curOrderFlat:
      trend='='
    else:
      if slopeRand<(curOrderAvg):
        trend='-'
      else:
        trend='=' 
  elif (initTrend=='+'):
   if trendRand<=curOrderFlat:
     trend='='
   else:
     if slopeRand>(curOrderAvg):
       trend='+'
     else:
       trend='='  
  steps=[]
  if trend!='=':
    targetGrade=curOrderMax+1.0
    while targetGrade>curOrderMax:
      targetGrade=abs(random.gauss(0.0,(curOrderMax/3.6)))
# step 2: determine number of chunks  
    chunksQty=1+int(6.0*(targetGrade/curOrderMax))
# step 3: determine grade change per chunk 
    if trend=='-':
      step=-1.0*(targetGrade/chunksQty)
    else:
      step=targetGrade/chunksQty
    curGrade=0.0
#    print "target: "+str(targetGrade)
#    print "steps: "+str(chunksQty)
    for i in range (0,chunksQty):
      curGrade=curGrade+step
      steps.append(curGrade)
#    print steps
  else:
    steps.append(0.0)
# step 4: determine chunks length
#  print steps
  lens=[]
  for s in steps:
    len0=MAXSECTION+1.0
    while len0>MAXSECTION:
      len0=MINSECTION*longTail(1.4,0.5,MAXSECTION)
    gradient=int(s*10000.0)
    gradient=gradient/100.0
    cleanX=float(int(x/1.0))/1000.0
    cleanZ=float(int(z*100.0))/100.0
#    print "pK:"+str(x/1000.0)+" gradient:"+str(gradient)+"("+str(s)+")"+" "+" length:"+str(len0)+" "+str(z)
    if ((gradient!=oldGradient) or (initTrend=='0')):
#      print str(x/1000.0)+" "+str(gradient)+" "+str(z)
      print str(33.9+cleanX)+" "+str(gradient)+" "+str(cleanZ)
    z=z+s*len0
    if ((x+len0)<lenM):
      x=x+len0
    oldGradient=gradient
  return trend

def longTail(startpoint,incr,maxval):
    if (startpoint<1.0):
      return -2 # not fat enough!
    returnval=1
    point=startpoint
    h=random.uniform(0.0,1.0)
    while True:
      if (returnval>maxval):
        return maxval #retry with another thincounter
      if (h<(1/point)):
        point=point+incr
        returnval=returnval+1
      else:
        return float(returnval)*random.uniform(1.0,2.0)


lenM=LENGTH*1000.0
oldTrend="0"
orderCnt=1
nextOrderX=float(secondOrder[orderCnt]['x'])
curOrderAvg=float(secondOrder[orderCnt-1]['avg'])
curOrderFlat=float(secondOrder[orderCnt-1]['flat'])
curOrderMax=float(secondOrder[orderCnt-1]['max'])
while (x<(lenM-MINSECTION)):
  trend=section(oldTrend)
  oldTrend=trend
  if nextOrderX<x:
#    print "==== second order ===="
    orderCnt=orderCnt+1
    nextOrderX=float(secondOrder[orderCnt]['x'])
    curOrderAvg=float(secondOrder[orderCnt-1]['avg'])
    curOrderFlat=float(secondOrder[orderCnt-1]['flat'])
    curOrderMax=float(secondOrder[orderCnt-1]['max'])
print str(33.9+LENGTH)+" "+str(0.0)+" "+str(float(int(z*100.0))/100.0)


tools/profileReverse.py
#!/usr/bin/python

import sys

sourceProfile=[]
reversedProfile=[]

def loadProfile():
  f=open("GRDs.txt","r")
  ssf=f.readlines()
  ss=[]
  f.close()
  cnt=0
  for s in ssf:
    if (s[0]!='#'):
      s=s.rstrip()
      ss.append(s)
      cnt=cnt+1
  return ss

def reverse():
  global sourceProfile
  global reversedProfile
  sourceProfile=loadProfile()
  lastRow=sourceProfile[-1]
  lenkm=float(lastRow.split(" ")[0])
  cntLines=0
  for p in sourceProfile:
    prev=[]
    prev=[0.0,0.0,0.0]
    if cntLines>0:
      oldRow=curRow
    curRow=p.split(" ")
    if cntLines>0:
      prev[0]=lenkm-float(curRow[0])
      if oldRow[1]=='0.0':
        prev[1]=0.0
      else:
        prev[1]=-1.0*float(oldRow[1])
      #prev[2]=float(curRow[2])
    else:
      prev[0]=lenkm-float(curRow[0])
      prev[1]=0.0
      #prev[2]=float(curRow[2])
    #print str(lenkm-curX)+" "+str(curGRD)+" "+str(curZ)
    reversedProfile.append(prev)
    cntLines=cntLines+1
  for p in reversed(reversedProfile):
    print str(p[0])+" "+str(p[1])+" "+str(p[2])

reverse()


tr.py
#!/usr/bin/python

import re,sys,redis,json,getopt,random,urllib2
import time,uuid
MPH=1.60934
WAKEINTERVAL=3

alreadyPicked=[]
liveScheds=[]
maps=[]
segMaps=[]
projectDir=""
segmentsDir="segments/"
foundSch=False
conf={}
#scheduleFile=""
redisDB=0

try:
  opts, args = getopt.getopt(sys.argv[1:], "h:m", ["help", "instance="])
except getopt.GetoptError as err:
  print(err) # will print something like "option -a not recognized"
  usage()
  sys.exit(2)

for o, a in opts:
  if o in ("--instance"):
    redisDB = int(a)
#  elif o in ("--schedule"):
#    foundSch=True
#    scheduleFile = a
  else:
    assert False, "option unknown"
    sys.exit(2)
#if foundSch==False:
#  print "ERROR. Schedule needed."
#  sys.exit(2)

r=redis.StrictRedis(host='localhost', port=6379, db=redisDB)
try:
  answ=r.client_list()
except redis.ConnectionError:
  print "FATAL: cannot connect to redis."
  sys.exit()

routeName=r.get("routeName")
projectDir=routeName+'/'

def getOffset(seg,name):
  global unitsFactor
  f=open(projectDir+"segments/"+seg+"/SIGs.txt","r")
  ssf=f.readlines()
  ss=[]
  f.close()
  cnt=0
  for s in ssf:
    s=s.split(" ")
    if s[1]==name:
      return float(s[0])*1000.0*unitsFactor
  return None

def getSegLenAndType(seg):
  global unitsFactor
  f=open(projectDir+"segments/"+seg+"/SIGs.txt","r")
  ssf=f.readlines()
  ss=[]
  f.close()
  cnt=0
  for s in ssf:
    if cnt==0:
      s=s.split(" ")
      if len(s)>2:
        firstS=s[2]
      else:
        firstS='1'
      firstN=s[1]
      if len(s)>3:
        firstB=s[3]
    cnt=cnt+1
  s=s.split(" ")
  if len(s)>2:
    lastS=s[2]
  else:
    lastS='1'
  lastN=s[1]
#  print seg+" "+str(firstS)+"->"+str(lastS)
  lnt={}
  lnt['length']=1000.0*float(s[0])*unitsFactor
  typeFound=False
  if firstS=='2':
    if lastS=='2':
      typeFound=True
      lnt['type']='Main'
      lnt['offset']=0.0
    elif lastS=='4C':
      typeFound=True
      lnt['type']='BranchConv'   
      segMain=s[3].split(":") 
      if segMain[0]=='Main':
        lnt['offset']=getOffset(segMain[1],lastN)
      else:
        print "FATAL: syntax error in signal "+lastS+":"+lastN
        sys.exit()
  elif firstS=='4D':
    if lastS=='2':
      typeFound=True
      lnt['type']='BranchDiv'
      segMain=firstB.split(":") 
      if segMain[0]=='Main':
        lnt['offset']=getOffset(segMain[1],firstN)
      else:
        print "FATAL: syntax error in signal "+firstS+":"+firstN
        sys.exit()
    elif lastS=='1':
      typeFound=True
      lnt['type']='GarageExit'
      lnt['offset']=0.0
    elif lastS=='4C':
      typeFound=True
      lnt['type']='Siding'
      lnt['offset']=0.0
  elif firstS=='1':
    if lastS=='4C':
      typeFound=True
      lnt['type']='GarageFeed'
      lnt['offset']=0.0
    elif lastS=='1':
      typeFound=True
      lnt['type']='OneOff'
      lnt['offset']=0.0
  if 'type' in lnt:
    return lnt
  print "FATAL: unknown segment type: "+seg+" firstS:"+firstS+" lastS:"+lastS
  sys.exit()
  return None

def confMAPs():
  f=open(projectDir+"maps.txt","r")
  ssf=f.readlines()
  ss=[]
  f.close()
  cnt=0
  aL=[]
  for s in ssf:
    if s[0]!='#':
      s=s.rstrip().split(":")
      aS={}
      ss=s[1].split(",")
      aS[s[0]]=ss
      aL.append(aS)
  return aL

def initSEGmap(sm):
  global segMaps
  global segmentsList
  global liveScheds
  schml=[]
  schm={}
  for ll in liveScheds:
    if ll['segment']==sm:
      lli={}
      lli['x']=ll['x']
      lli['name']=ll['name']
      schml.append(lli)
  schm['points']=schml
  lnt=getSegLenAndType(sm)
  schm['length']=lnt['length']
  schm['type']=lnt['type']
  schm['offset']=lnt['offset']
#  print schm
  return schm

def initMAPs():
  global maps
  global segMaps
  global segmentsList
# stage 1: identify maps
  mapsconf=confMAPs()
#  print mapsconf
# stage 2: generate segment maps
  for se in segmentsList:
    found=False
    for sm in segMaps:
      if sm['segment']==se:
        found=True
    if found==False:
      smp={}
      smp['map']=initSEGmap(se)
      smp['segment']=se
      segMaps.append(smp)
#  print segMaps
#stage 3: generate maps
  lmaps=[]
  for m in mapsconf:
    hasMain=False
    keys=m.keys()
    if keys[0] is not None:
      cnt=0
      amap={}
      amap['map']=keys[0] 
      amap['points']=[]
      for mm in m[keys[0]]:
        for sm in segMaps:
          if sm['segment']==mm:
            if ((hasMain==False) and (sm['map']['type']=='Main')):
              hasMain=True
#              print "MAIN"
              for p in sm['map']['points']:
                amap['points'].append(p)
#                print amap['points']
#              print "/MAIN"
            elif ((hasMain==True) and (sm['map']['type']=='Main')):
              print "FATAL: map has two main branches"
              sys.exit()
            elif sm['map']['type']=='BranchDiv':
#              print "BDIV"
              for p in sm['map']['points']:          
                p1={}
                p1['x']=p['x']+sm['map']['offset']
                p1['name']=p['name']
                amap['points'].append(p1)
#              print "/BDIV"
            elif sm['map']['type']=='BranchConv':
#              print "BCONV"
              for p in sm['map']['points']:          
                p1={}
                p1['x']=p['x']+sm['map']['offset']-sm['map']['length']
                p1['name']=p['name']
                amap['points'].append(p1)
#              print "/BCONV"
#      print 'before lambda:'
#      print amap['points']
      amap['points'].sort(key=lambda l: l['x'])
#      print 'after lambda:'
#      print amap['points']
      lmaps.append(amap)
#  print "LMAPS:"
#  print lmaps
#  print "/LMAPS"
  return lmaps

def distances(m):
  global alreadyPicked
  oldp={}
  dl=[]
  for p in m['points']:
    if 'name' in oldp:
      found=False
      for a in alreadyPicked:
        if a==oldp['name']:
          found=True
      if found==False:
        d={}
        d['distance']=p['x']-oldp['x']
        d['from']=oldp['name']
        d['to']=p['name']
#      print d
        dl.append(d)
    oldp=p
  dl.sort(key=lambda l: l['distance'])
  return dl

def initSRVs(s_type):
  f=open(projectDir+"services.txt","r")
  ssf=f.readlines()
  ss=[]
  f.close()
  cnt=0
  if s_type=='exit':
    s_char='-'
  elif s_type=='standard':
    s_char='+'
  else:
    return []
  a=[]
  for s in ssf:
    aS={}
    if (s[0]==s_char):
      s=s.rstrip().split(" ")
      aS['name']=s[1]
      aDivs=[]
      if s_char=='+':
        ss=s[2].split(":")
        if ss[0]!='TpH':
          print "FATAL: unknown service verb"
          sys.exit()
        ss=ss[1].split("-")
        aS['min']=int(ss[0])
        aS['max']=int(ss[1])
      elif s_char=='-':
        if s[2]!='OneOff':
          print "FATAL: unknown service verb"
          sys.exit()
      for i in range(3,len(s)):
#        print "s"+str(i)+": "+s[i]
        ss=s[i].split(":")
        aSS={}
        aSS[ss[0]]=ss[1]
        aDivs.append(aSS) 
      aS['divs']=aDivs
      cnt=cnt+1
      a.append(aS) 
  return a

def initSEGs():
  f=open(projectDir+"segments.txt","r")
  ssf=f.readlines()
  ss=[]
  f.close()
  cnt=0
  for s in ssf:
    if (s[0]!='#'):
      s=s.rstrip()
      ss.append(s)
      cnt=cnt+1
  return ss

def initConfig():
  f=open(projectDir+"routeConfig.txt","r")
  ssf=f.readlines()
  ss=[]
  f.close()
  cnt=0
  for s in ssf:
    if (s[0]!='#'):
      s=s.rstrip().split(" ")
      ss.append(s)
      cnt=cnt+1
  return ss

def initSchedules(seg,sFile):
  global segs
  f=open(projectDir+"/schedules/"+sFile,"r")
  ssf=f.readlines()
  ss=[]
  f.close()
  cnt=0
  for s in ssf:
    if (s[0]!='#'):
      s=s.rstrip().split(" ")  
      ss.append(s[0])
      cnt=cnt+1
  return ss

def initSTAs(seg):
  global segs
  f=open(projectDir+segmentsDir+seg+"/STAs.txt","r")
  ssf=f.readlines()
  ss=[]
  f.close()
  cnt=0
  for s in ssf:
    if (s[0]!='#'):
      s=s.rstrip().split(" ")
      ss.append(s)
      cnt=cnt+1
  return ss

def initSIGs(seg):
  global segs
  f=open(projectDir+segmentsDir+seg+"/SIGs.txt","r")
  ssf=f.readlines()
  ss=[]
  f.close()
  cnt=0
  for s in ssf:
    if (s[0]!='#'):
      s=s.rstrip().split(" ")
      ss.append(s)
      cnt=cnt+1
  return ss

def getState():
  global liveScheds
  ll=[]
  rs=urllib2.urlopen('http://127.0.0.1:4999/v1/list/schedules')
  sched=json.loads(rs.read())
  for v in sched:
    state=r.hgetall('state:'+v)
    state=dict(state)
    s1={}
    if 'segment' in state:
      s1['name']=v
      s1['inSta']=state['inSta']
      s1['atSig']=state['atSig']
      s1['coasting']=state['coasting']
      s1['sigBrake']=state['sigBrake']
      s1['staBrake']=state['staBrake']
      s1['segment']=state['segment']
      s1['nextSTA']=state['nextSTA']
      s1['nextSIG']=state['nextSIG']
      s1['nextTIV']=state['nextTIV']
      s1['service']=state['service']
      s1['advSIGcol']=state['advSIGcol']
      s1['x']=float(state['x'])
      s1['v']=float(state['v'])
      s1['pax']=float(state['pax'])
      s1['maxPax']=float(state['maxPax'])
      s1['maxVk']=float(state['maxVk'])
      s1['units']=state['units']
      ll.append(s1)
  return ll

def getSegExits(seg,sig):
  sis=initSIGs(seg)
  exits=[]
  cntEx=0
  exPos=-1
  for si in sis:
    if len(si)>4:
      if si[2]=='4D':
        br=si[4].split(":")
        if br[0]=='Branch':
          chk=getSegLenAndType(br[1])
          if 'type' in chk:
            if chk['type']=='GarageExit':
              order={}
              order['type']=si[2]
              order['name']=si[1]
              bro=si[5].split(":")
              if bro[0]=='BranchOrientation':
                order['direction']=bro[1]
              exPos=cntEx
              cnt=0    
              sigPos=-1
              for si2 in sis:
                if si2[1]==sig:
                  sigPos=cnt
                cnt=cnt+1
              if sigPos>0:
#                print "sch sig pos is:"+str(sigPos)
#                print "div sig pos is:"+str(exPos)
                if ((sigPos>exPos) or (sigPos<(exPos-2))):  #train must not be too close to div
                  exits.append(order)
    cntEx=cntEx+1
  return exits

def getReverseSeg(seg):
  chk=getSegLenAndType(seg)
  if 'type' in chk:
    if chk['type']=='Main':
      sis=initSIGs(seg)
      si=sis[-1][3].split(":")
      if si[0]=='Reverse':
        return si[1]
  return None

def getExitsForSched(sname):
  global liveScheds 
  for l in liveScheds:
    if l['name']==sname:
      break
  if l is not None:
    sseg=l['segment']
#    print sname+" is in seg: "+sseg
    exits=getSegExits(sseg,l['nextSIG'])
    if len(exits)>0:
      return exits
    else:
      altSeg=getReverseSeg(sseg)
      if altSeg is not None:
        exits=getSegExits(altSeg,l['nextSIG'])
        if len(exits)>0:
          return exits
  return None

segmentsList=initSEGs()
confraw=initConfig()
unitsFactor=1.0

for aa in confraw:
  if (aa[0]!="#"):
    if (aa[0]=='units'):
      conf['units']=aa[1]
      if conf['units']=='imperial':
        unitsFactor=MPH

downScaleQty=4
candidatesQty=downScaleQty*2
liveScheds=getState()
candidatesQty=min(candidatesQty,len(liveScheds))
candidatesQty=max(candidatesQty,8)
selected=[]
maps=initMAPs()
simID=r.get("simID")
#exits=r.lrange(simID+":exits",0,-1)
loopPick=True
cnt=0
scalingEvent={}
scalingEvent['type']='downScale'
scalingEvent['desired']=downScaleQty
scalingEvent['id']=str(uuid.uuid4())
scalingEvent['time']=r.get("elapsed")
scalingEvent['simID']=simID
while loopPick==True:
#  print "ITERATION "+str(cnt)
  for m in maps:
    dl=distances(m)
#    print dl
    if candidatesQty>0:
      alreadyPicked.append(dl[0]['from']) 
      candidatesQty=candidatesQty-1
    else:
      loopPick=False
  cnt=cnt+1

print alreadyPicked

cnt=0
for a in alreadyPicked:
  exits=getExitsForSched(a)
  if exits is not None:
    cnt=cnt+1
    candidate={}
    candidate['name']=a
#    candidate['order']=exits[0]['name']+':'+exits[0]['direction']
    candidate['sigName']=exits[0]['name']
    candidate['sigDir']=exits[0]['direction']
    if cnt<=downScaleQty:
      selected.append(candidate)
print selected
scalingEvent['set']=len(selected)
print scalingEvent

for e in selected:
  rs=urllib2.urlopen('http://127.0.0.1:4999/v1/update/schedule/'+e['name']+'/'+e['sigName']+'/'+e['sigDir']+'/'+scalingEvent['time'])
  sched=json.loads(rs.read())
  print "RES:"
  print sched

def mainLoop():
  global liveScheds
  global elapsed
  global srvs
  while True:
    elapsed=r.get("elapsed")
    if elapsed is not None:
      elapsed=int(elapsed)
      liveScheds=getState()
      print elapsed
    else:
       print "lost connection to redis...retrying..."
    time.sleep(WAKEINTERVAL)

#mainLoop()
